APIMatic.Core.Test/AuthenticationTest.cs

```diff
@@ -0,0 +1,137 @@
+    {
+        [Test]
+        public void Multiple_Authentication_Success()
+        {
+            var basicAuthManager = new BasicAuthManager("username", "password");
+            var globalConfiguration = new GlobalConfiguration.Builder()
+                .ServerUrls(new Dictionary<Enum, string>
+                {
+                    {MockServer.Server1, "http://my/path:3000/{one}"},
+                }, MockServer.Server1)
+                .AuthManagers(new Dictionary<string, AuthManager>()
+                {
+                    {"basic", basicAuthManager},
+                    {"header", new HeaderAuthManager("my api key", "test token")},
+                    {"query", new QueryAuthManager("my api key", "test token")}
+                })
+                .HttpConfiguration(_clientConfiguration)
+                .Build();
+
+            var request = globalConfiguration.GlobalRequestBuilder()
+                .Setup(HttpMethod.Get, "/auth")
+                .WithOrAuth(auth => auth
+                    .Add("basic")
+                    .AddAndGroup(innerGroup => innerGroup
+                        .Add("header")
+                        .Add("query")))
+                .Build();
+
+            Assert.AreEqual(basicAuthManager.GetBasicAuthHeader(), request.Headers["Authorization"]);
+            Assert.AreEqual("my api key", request.Headers["API-KEY"]);
+            Assert.AreEqual("test token", request.Headers["TOKEN"]);
+            Assert.AreEqual("my api key", request.QueryParameters["API-KEY"]);
+            Assert.AreEqual("test token", request.QueryParameters["TOKEN"]);
+        }
+
+        [Test]
+        public void Multiple_Authentication_OR_Validation_Failure()
+        {
+            var globalConfiguration = new GlobalConfiguration.Builder()
+                .ServerUrls(new Dictionary<Enum, string>
+                {
+                    {MockServer.Server1, "http://my/path:3000/{one}"},
+                }, MockServer.Server1)
+                .AuthManagers(new Dictionary<string, AuthManager>()
+                {
+                    {"basic", new BasicAuthManager("username", null)},
+                    {"header", new HeaderAuthManager("my api key", null)},
+                    {"query", new QueryAuthManager("my api key", "test token")}
+                })
+                .HttpConfiguration(_clientConfiguration)
+                .Build();
+
+            var exp = Assert.Throws<ArgumentNullException>(() => globalConfiguration.GlobalRequestBuilder()
+                .Setup(HttpMethod.Get, "/auth")
+                .WithOrAuth(auth => auth
+                    .Add("basic")
+                    .AddAndGroup(innerGroup => innerGroup
+                        .Add("header")
+                        .Add("query")))
+                .Build());
+
+            Assert.AreEqual("Following authentication credentials are required:\n" +
+                "-> Missing required header field: Authorization\n" +
```

Comment:
Question: Is there a way in your implementation that shows the auth implementation where these credentials are missing? For instance, in this particular case, error message should point to Basic Auth and Header Auth etc.

Reply:
No, That's not possible. The name of the exact auth parameter is the only way we can pinpoint the missing required field.

However, the actual auth managers are not exposed to SDK Users. An SDK user should only be concerned with the auth parameter name. That's what we show in the ReadMe for client configurations section